gdx/src/com/badlogic/gdx/utils/Pool.java

```diff
@@ -48,19 +48,32 @@ public Pool (int initialCapacity, int max) {
 	/** Returns an object from this pool. The object may be new (from {@link #newObject()}) or reused (previously
 	 * {@link #free(Object) freed}). */
 	public T obtain () {
-		return freeObjects.size == 0 ? newObject() : freeObjects.pop();
+                if (freeObjects.size == 0) {
+                        return newObject();
+                } else {
+                        T object = freeObjects.pop();
+                        onObtain(object);
+                        return object;
+                }
 	}
-
+	
+	/** Called when an object is retrieved from the pool using {@link #obtain()} */
+	protected void onObtain (T object){ }

```

Comment:
The documentation is not 100% accurate: object may be returned by the pool, but without calling `onObtain` - if it is created by the Pool. I don't think that this is necessarily a problem, but it should be documented as it will lead to surprises. Same problem at `onFree`.

Looking at the whole PR, it feels that it would be most beneficial:
- if the `onObtain`/`onFree` were called deterministically (i.e. as documented, not as implemented),
- whole onObtain/onFree could be then implemented directly in `DescriptivePool`, without any changes to `Pool` necessary,
- and the addition of `onFree`/`onRelease` for overriding directly in the `Pool` subclass seems, in retrospect, much more useful than generic bounds.

All ideas are open to discussion, sorry for not realizing these sooner :/

Reply:
The reason I made it that way for `onObtain`/`onFree` to be called only when actually retrieved or put back is because with `onObtain`, the object's constructor is simply called and I didn't want those two logics to be mixed, worse case scenario you can call `onObtain` from within the object's constructor if you want. So it allows for more configurability-- I guess. 

I'm not entirely sure what you're suggesting as far as your last bullet point. Are you saying remove `DescriptivePoolable` and just have the methods `onFree`/`onObtain` defined in `DescriptivePool` and have the implementation details open to subclassing `DescriptivePool`?

Reply:
Sorry for not being clearer. What I meant was, that when `Pool` implementation itself has `onObtain`/`onFree`, requiring the pooled objects to implement `DescriptivePoolable` is superfluous, because the implementation of whatever should happen `onObtain`/`onFree` can be implemented directly in the `Pool` subclass. (This would, in turn, mean, that default `onObtain`/`onFree` implementation in the `Pool` must do `instanceof` checks, which is a small price to pay for that flexibility I guess.)

And with the first part, I agree with you that not calling `onObtain` when the object is created may be beneficial, I think that this does not hold for `onFree`, because user may want to implement, for example, native resource disposal there. Then the game would work as long as the `Pool` operates within limits, but as soon as it starts throwing freed objects away, the game will leak resources, in a non-obvious way.

And if `onFree` is called always, it feels right (this is subjective, I know) to have it symmetrical with `onObtain`. (I suspect that many types of pooled objects won't need any explicit constructor at all in that case.) And when both `onFree` and `onObtain` are called always, it is possible to implement that functionality in `DescriptivePool` without changing `Pool` itself, which is a nice benefit, because then there is no need to worry about breaking it.

In the end, those are just my opinions and the final decision is on you and the owners.

Reply:
Hmmmmmmm..... I see where you're coming from and I guess you're quite right. I agree now with onObtain and onFree. Lemme make those changes